webapp/scripts/modules/query.js

```diff
@@ -0,0 +1,183 @@
+const fs = require('fs')
+const { addVector, dbSelectAll, getAllColumns, getLayers, getUnivarBounds, gpkgOut, grass, initMapset, listUserVector, mapsetExists, remove } = require('../grass')
+const { checkWritableDir, filterDefaultLayers, getFilesOfType, mergePDFs, psToPDF, textToPS } = require('../helpers')
+const { query: messages } = require('../messages.json')
+
+const GEOSERVER = `${process.env.GEOSERVER_DATA_DIR}/data`
+const GRASS = process.env.GRASS_DIR
+const OUTPUT = process.env.OUTPUT_DIR
+
+const QUERY_RESOLUTION = 0.00002
+const QUERY_MAP_NAME = 'query_map'
+const QUERY_RESULT_NAME = 'query_result'
+
+module.exports = class {
+  constructor() {
+    this.mapset = 'query'
+  }
+
+  launch() {
+    checkWritableDir(GEOSERVER)
+    checkWritableDir(OUTPUT)
+
+    if (!mapsetExists('PERMANENT')) {
+      return messages[7]
+    }
+
+    initMapset(this.mapset)
+
+    // set selection as the query area
+    grass(this.mapset, `g.copy vector=selection@PERMANENT,selection --overwrite`)
+    this.queryArea = 'selection'
+
+    const allVector = listUserVector()
+    const allGpkg = getFilesOfType('gpkg', GEOSERVER).filter(filterDefaultLayers)
+
+    /**
+     * check and add all layers from layer switcher
+     * FIXME: mapFile name and layer name differs, which makes it hard to check if file has already been imported
+     * ideal solution: 1. make sure all map files contains only one layer; 2. make sure all mapFile names align with its layer's name
+     */
+    for (const mapFile of allGpkg) {
+      const fileName = mapFile.slice(mapFile.lastIndexOf('/') + 1, mapFile.lastIndexOf('.'))
+      if (allVector.indexOf(fileName) > -1) continue
+
+      const inLayers = getLayers('PERMANENT', mapFile)
+      // GRASS doesn't allow space in layernames
```

Comment:
Is this the only limitation that applies?

Reply:
I wrote a letter to ask Titusz about this.

Reply:
https://grass.osgeo.org/grass78/manuals/libpython/_modules/script/core.html#legal_name